lib/compat/wordpress-6.6/blocks.php

```diff
@@ -44,3 +44,237 @@ function gutenberg_replace_pattern_override_default_binding( $parsed_block ) {
+	// Create private anonymous class until the HTML API provides `set_inner_html` method.
+	$block_reader = new class($block_content) extends WP_HTML_Tag_Processor{
+		/**
+		 * THESE METHODS ARE A TEMPORARY SOLUTION NOT TO BE EMULATED.
+		 * IT IS A TEMPORARY SOLUTION THAT JUST WORKS FOR THIS SPECIFIC
+		 * USE CASE UNTIL THE HTML PROCESSOR PROVIDES ITS OWN METHOD.
+		 */
+
+		/**
+		 * Replace the inner content of a figcaption element with the passed content.
+		 *
+		 * @param string $new_content New content to insert in the figcaption element.
+		 * @return bool Whether the inner content was properly replaced.
+		 */
+		public function gutenberg_set_content_between_figcaption_balanced_tags( $new_content ) {
+			// Check that the processor is paused on an opener tag.
+			if ( $this->is_tag_closer() || 'FIGCAPTION' !== $this->get_tag() ) {
+				return false;
+			}
+
+			// Set position of the opener tag.
+			$this->set_bookmark( 'opener_tag' );
+
+			/*
+			 * This is a best-effort guess to visit the closer tag and check it exists.
+			 * In the future, this code should rely on the HTML Processor for this kind of operation.
+			 */
+			$tag_name = $this->get_tag();
+			if ( ! $this->next_tag(
+				array(
+					'tag_name'    => $tag_name,
+					'tag_closers' => 'visit',
+				)
+			) || ! $this->is_tag_closer() ) {
+				return false;
+			}
+
+			// Set position of the closer tag.
+			$this->set_bookmark( 'closer_tag' );
+
+			// Get opener and closer tag bookmarks.
+			$opener_tag_bookmark = $this->bookmarks['opener_tag'];
+			$closer_tag_bookmark = $this->bookmarks['closer_tag'];
+
+			// Appends the new content.
+			// Compat for 6.4, where bookmarks and WP_HTML_Text_Replacement are different.
+			if ( ! empty( $opener_tag_bookmark->end ) ) {
+				$this->lexical_updates[] = new WP_HTML_Text_Replacement( $opener_tag_bookmark->end + 1, $closer_tag_bookmark->start, $new_content );
+				return true;
+			}
+
+			$after_opener_tag = $opener_tag_bookmark->start + $opener_tag_bookmark->length;
+			/*
+			 * There was a bug in the HTML Processor token length fixed after 6.5.
+			 * This check is needed to add compatibility for that.
+			 * Related issue: https://github.com/WordPress/wordpress-develop/pull/6625
+			 */
+			if ( '>' === $this->html[ $after_opener_tag ] ) {
+				++$after_opener_tag;
+			}
+			$inner_content_length    = $closer_tag_bookmark->start - $after_opener_tag;
+			$this->lexical_updates[] = new WP_HTML_Text_Replacement( $after_opener_tag, $inner_content_length, $new_content );
+			return true;
+		}
+
+		/**
+		 * Add a new figcaption element after the `img` or `a` tag.
+		 *
+		 * @param string $new_element New figcaption element to append after the tag.
+		 * @return bool Whether the element was properly appended.
+		 */
+		public function gutenberg_append_figcaption_element_after_tag( $new_element ) {
+			$tag_name = $this->get_tag();
```

Comment:
Should we escape `$new_element` here?

We are passing `<figcaption html>$attributes['caption']</figcaption>`.